internal/codecs/object.go

```diff
@@ -53,19 +59,33 @@ func buildObjectDecoder(
 			sf.Type,
 			path.AddField(field.Name),
 		)
-
 		if err != nil {
 			return nil, err
 		}
 
+		if !field.Required {
+			if _, isOptional := child.(OptionalDecoder); !isOptional {
+				return nil, fmt.Errorf(
+					"expected %v at %v.%v to implement OptionalUnmarshaler "+
```

Comment:
```suggestion
					"expected %v at %v.%v to implement OptionalDecoder "+
```

I'd also replace (un)marshaller with (de)coder everywhere to be consistent.

Reply:
I thought about using Decoder. I went with Unmarshaller instead because it is what the stdlib uses https://pkg.go.dev/encoding

Reply:
Why is the interface named `OptionalDecoder` then?

Reply:
Probably because `Decoder` is used in `internal/` types and I mixed them up when I wrote this. I think the interface should be `OptionalUnmarshaller`.

Reply:
It looks like I actually have both an `OptionalUnmarshaler` and `OptionalDecoder` interface and they have different methods. Let me thing about this and push a change :thinking:

Reply:
On looking at this again, `type OptionalDecoder interface` is an internal API that is implemented by Decoder types defined in `internal/`. The suggested change is correct. I have updated the PR.

> I'd also replace (un)marshaller with (de)coder everywhere to be consistent.

I think that (Un)Marshaler is a more helpful name for the public API because of its use in the stdlib.

Reply:
The type cast to `OptionalDecoder` is checking if the internal decoder  knows how to treat the user's out value as optional. If not the error message reports to the user that their out type needs to implement `OptionalUnmarshaler`. This error message is correct.

Reply:
Here is the assertion that tests this behavior https://github.com/edgedb/edgedb-go/pull/136/files#diff-02776af2e303aa4b13cf883561bfd1aa91a940389c39aae1b7abbaecc0d68092R3416